framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/OpenShiftDeployment.java

```diff
@@ -88,38 +101,85 @@ public boolean isReady() {
-            List<Pod> pods = openShiftController.getClient().pods().inNamespace(namespace).withLabel(OpenShiftResourceConstants.DEPLOYMENT_CONFIG_LABEL, deploymentConfigName).list().getItems();
+            // Deployment config has a same name as its service.
+            String deploymentConfigName = getServiceName();
 
-            List<Instance> instances = pods.stream().map((pod) -> {
-                return createInstance(pod);
-            }).collect(toList());
+            List<Instance> instances = util.getPods().stream()
+                    .filter(pod -> pod.getMetadata().getLabels().get(OpenShiftResourceConstants.DEPLOYMENT_CONFIG_LABEL) != null
+                                && pod.getMetadata().getLabels().get(OpenShiftResourceConstants.DEPLOYMENT_CONFIG_LABEL).equals(deploymentConfigName))
+                    .map(pod -> createInstance(pod))
+                    .collect(toList());
 
             return instances;
         }
 
         return Collections.emptyList();
     }
 
+    @Override
+    public void waitForScale() {
+        waitUntilAllPodsAreStarted();
+        waitUntilAllPodsAreRunning();
+    }
+
+    private void waitUntilAllPodsAreStarted() {
+        int expectedPods = util.getDeploymentConfig(getServiceName()).getSpec().getReplicas().intValue();
+
+        BooleanSupplier arePodsReady = () -> {
+            List<Pod> pods = util.getLabeledPods(Collections.singletonMap(OpenShiftResourceConstants.DEPLOYMENT_CONFIG_LABEL, getServiceName()));
+
+            if (pods.size() == expectedPods) {
+                return pods.stream().allMatch(WaitUtil::isPodReady);
+            } else {
+                return false;
+            }
+        };
+
+        try {
+            WaitUtil.waitFor(arePodsReady, null, WaitUtil.DEFAULT_WAIT_INTERVAL, OpenShiftResourceConstants.PODS_START_TO_READY_TIMEOUT);
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Exception while waiting for scale.");
+        } catch (TimeoutException e) {
+            throw new DeploymentTimeoutException("Timeout while waiting for pods to start.");
+        }
+    }
+
+    private void waitUntilAllPodsAreRunning() {
+        BooleanSupplier arePodsRunning = () -> {
+            List<Pod> pods = util.getLabeledPods(Collections.singletonMap(OpenShiftResourceConstants.DEPLOYMENT_CONFIG_LABEL, getServiceName()));
+            return pods.stream().allMatch(WaitUtil::isPodRunning);
+        };
+
+        try {
+            WaitUtil.waitFor(arePodsRunning, null, WaitUtil.DEFAULT_WAIT_INTERVAL, OpenShiftResourceConstants.PODS_START_TO_READY_TIMEOUT);
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Exception while waiting for scale.");
+        } catch (TimeoutException e) {
+            throw new DeploymentTimeoutException("Timeout while waiting for pods to start.");
```

Comment:
You have the same DeploymentTimeoutException message in waitUntilAllPodsAreRunning and in waitUntilAllPodsAreStarted. Is that intended?
Also can you explain what's the difference between pods started and pods ready?

Reply:
Yes, it is intended. The DeploymentTimeoutException marks some problem with pod deployment which prevented pod to start. It may be the case that pod wasn't initialized correctly and therefore is unable to start, or pod is initialized but failed to start.
Method waitUntilAllPodsAreStarted was renamed to waitUntilAllPodsAreReady to better describe its functionality. It checks that all pods are available and ready to be started (or already starting).
Ready pod means that pod is initialized and ready to be started. Running pod means that pod is ready to server requests - passed readiness probe.